core/node/node_framework/src/task.rs

```diff
@@ -23,18 +55,97 @@ pub trait Task: 'static + Send {
+pub trait OneshotTask: 'static + Send {
+    /// Unique name of the task.
+    fn name(&self) -> &'static str;
+
+    /// Runs the task.
     ///
-    /// This hook can be used to perform some cleanup that assumes exclusive access to the resources, e.g.
-    /// to rollback some state.
+    /// Unlike [`Task::run`], this method is expected to return once the task is finished, without causing the
+    /// node to shutdown.
     ///
-    /// *Note*: This hook **should not** be used to perform trivial task cleanup, e.g. to wait for the spawned
-    /// server to stop. By the time this hook is invoked, every component of the node is expected to stop. Not
-    /// following this rule may cause the tasks that properly implement this hook to malfunction.
-    fn after_node_shutdown(&self) -> Option<BoxFuture<'static, ()>> {
-        None
+    /// `stop_receiver` argument contains a channel receiver that will change its value once the node requests
+    /// a shutdown. Every task is expected to either await or periodically check the state of channel and stop
+    /// its execution once the channel is changed.
+    ///
+    /// Each task is expected to perform the required cleanup after receiving the stop signal.
+    async fn run_oneshot(self: Box<Self>, stop_receiver: StopReceiver) -> anyhow::Result<()>;
+}
+
+impl dyn OneshotTask {
+    /// An internal helper method that guards running the task with a tokio Barrier.
+    /// Used to make sure that the task is not started until all the preconditions are met.
+    pub(super) async fn run_oneshot_with_barrier(
+        self: Box<Self>,
+        mut stop_receiver: StopReceiver,
+        preconditions_barrier: Arc<Barrier>,
+    ) -> anyhow::Result<()> {
+        // Wait either for barrier to be lifted or for the stop signal to be received.
+        tokio::select! {
+            _ = preconditions_barrier.wait() => {
+                self.run_oneshot(stop_receiver).await
+            }
+            _ = stop_receiver.0.changed() => {
+                Ok(())
+            }
+        }
     }
 }
+
+/// A task implementation that is not constrained by preconditions.
+///
+/// This trait is used to define tasks that should start immediately after the wiring phase, without waiting for
+/// any preconditions to be met.
+///
+/// *Warning*. An unconstrained task may not be aware of the state of the node and is expected to cautiously check
+/// any invariants it may rely on.
+#[async_trait::async_trait]
+pub trait UnconstrainedTask: 'static + Send {
+    /// Unique name of the task.
+    fn name(&self) -> &'static str;
+
+    /// Runs the task without waiting for any precondition to be met.
+    async fn run_unconstrained(self: Box<Self>, stop_receiver: StopReceiver) -> anyhow::Result<()>;
+}
+
+/// An unconstrained analog of [`OneshotTask`].
+/// See [`UnconstrainedTask`] and [`OneshotTask`] for more details.
+#[async_trait::async_trait]
+pub trait UnconstrainedOneshotTask: 'static + Send {
```

Comment:
Maybe a dumb suggestion: Could it make sense to have the base trait for all task types (potentially including preconditions) and use marker traits for each of the type? Looks like we have quite a few of task traits, but their signatures are almost identical. Yet another suggestion: add `fn task_type(&self) -> TaskType` to the task trait, where `TaskType` is defined like

```rust
enum TaskType {
    Normal,
    Oneshot,
    Unconstrained { oneshot: bool },
}
```

Reply:
Yeah, it was a deliberate decision. Reasons I had in mind for that:
1. I have heard several times already that people are concerned regarding having complex concepts in the framework. So _probably_ 4 simple concepts are actually easier to grok than one customizable (e.g. you get to choose one trait and then it's simple from there).
2. In fact, most of the time people want to use `Task`. So having it as a small trait with only mandatory fields feels like a plus: people are not distracted from what they want to do.
3. Base traits make the code harder to read, and I want (all) people to be able to read the framework code. E.g. write code in a "stupid" way where stupidity means easiness of access.
4. ...ultimately, it doesn't add too much complexity to the framework. Just a bit of boilerplate.

Or tl;dr I just thought that "stupid" way is more friendly for the people who will use the framework.

But indeed, I've spent some time thinking about it and not 100% decided, so will be happy to discuss.